x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -520,152 +534,222 @@ public void invalidateRefreshToken(String refreshToken, ActionListener<Boolean>
             maybeStartTokenRemover();
             findTokenFromRefreshToken(refreshToken,
                     ActionListener.wrap(tuple -> {
-                        final String docId = tuple.v1().getHits().getAt(0).getId();
-                        final long docVersion = tuple.v1().getHits().getAt(0).getVersion();
-                        indexInvalidation(docId, Version.CURRENT, listener, tuple.v2(), "refresh_token", docVersion);
+                        final String docId = getTokenIdFromDocumentId(tuple.v1().getHits().getAt(0).getId());
+                        indexInvalidation(Collections.singletonList(docId), listener, tuple.v2(), "refresh_token", null);
                     }, listener::onFailure), new AtomicInteger(0));
         }
     }
 
     /**
-     * Performs the actual bwc invalidation of a token and then kicks off the new invalidation method
+     * Invalidate all access tokens and all refresh tokens of a given {@code realmName} and/or of a given
+     * {@code username} so that they may no longer be used
      *
-     * @param userToken            the token to invalidate
+     * @param realmName the realm of which the tokens should be invalidated
+     * @param username the username for which the tokens should be invalidated
+     * @param listener  the listener to notify upon completion
+     */
+    public void invalidateActiveTokensForRealmAndUser(@Nullable String realmName, @Nullable String username,
+                                                      ActionListener<TokensInvalidationResult> listener) {
+        ensureEnabled();
+        if (Strings.isNullOrEmpty(realmName) && Strings.isNullOrEmpty(username)) {
+            logger.trace("No realm name or username provided");
+            listener.onFailure(new IllegalArgumentException("realm name or username must be provided"));
+        } else {
+            if (Strings.isNullOrEmpty(realmName)) {
+                findActiveTokensForUser(username, ActionListener.wrap(tokenTuples -> {
```

Comment:
I completely spaced on this, but we should use `UpdateByQuery`. I think it is safe to address this in a refactoring followup.

Reply:
I'll sync with you to get a clear idea of what you mean and will address this in a follow up